File: 25_dp/02_longest_increasing_subsequence.rs
use std::cmp::max;

impl Solution {
    pub fn length_of_lis(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut lis: Vec<i32> = vec![1; n];

        for i in (0..n).rev() {
            for j in i + 1..n {
                if nums[i] < nums[j] {
                    lis[i] = max(lis[i], 1 + lis[j]);
                }
            }
        }

        *lis.iter().max().unwrap()
    }
}

File: 25_dp/02_longest_increasing_subsequence.go
import "slices"

func lengthOfLIS(nums []int) int {
	n := len(nums)
	lis := make([]int, n)
	for i := range n {
		lis[i] = 1
	}
	for i := n - 1; i >= 0; i-- {
		for j := i + 1; j < n; j++ {
			if nums[i] < nums[j] {
				lis[i] = max(lis[i], 1+lis[j])
			}
		}
	}
	return slices.Max(lis)
}